README.md
# goiardi cookbook

This cookbook provides resources to install [goiardi](http://goiardi.gl/),
[schob](https://github.com/ctdk/schob/) and a complementary
[serf](https://www.serf.io/) resource (which may in time find it's way to a
seperate serf cookbook).

## Requirements

Tested on:

- Cinc Client version 15
- Centos 7
- Debian 8
- Debian 9
- Ubuntu 18.04
- Ubuntu 16.04
- Amazon Linux 2

Being assembled using only platform agnostic methods, it _should_ work on other
unix-like systems, but mileage may vary.

## Resources

### goiardi\_install

Installs a precompiled binary from github.com, or a custom url. Full syntax
with default values:

```
goiardi_install 'goiardi' do
  instance_name resource_name # Defaults to the resource name, used to generate paths and sub-resources
  instance_path '/usr/sbin'   # Where to put the binary. /usr/sbin is FHS compliant, but /usr/local/sbin could be argued as better.
  version 'latest'            # must be either 'latest' or a git tag. Ignored if bin_url is set explicitely
  bin_url nil                 # Left nil, the url will be autoresolved based on `version`
  user resource_name          # File and service owner
  group resource_name         # File group
  private_key_path nil        # Path to the signing private key. Left `nil`, one will be generated automatically if `use-shovey` is set in the goiardi configuration
  manage_systemd_service true # Set to false if you're not on systemd or prefer to write your own unit
  manage_conf true            # Set to false if you're supplying your own configuration file. You'll need to set `options` accordingly
  conf {}                     # Configuration entries to merge/override the defaults. Use `nil` to negate a default entirely
  options nil                 # CLI flags used in the service definition. Left nil, will default to `-c /path/to/conf`
  serf_unit_name 'serf'       # Will be added as `name.service` to the systemd unit's "After" and "Requires". Ignored if `manage_systemd_service` is `false`
  action :create              # Supports :create, :remove
end
```

Leaving `version` at it's default of 'latest' will result in goiardi
auto-updating with each client run. You probably don't want that happening in
prod o.O

`conf` is a hash of [Goiardi configuration
directives](https://goiardi.readthedocs.io/en/latest/installation.html#configuration)
that is merged in with some sane defaults, intended to create a standalone and
persistent Goiardi installation. Default directives can be overriden with
either a new value or `nil` to have them ommited entirely.

`options` is a string of [Goiardi CLI
options](https://goiardi.readthedocs.io/en/latest/installation.html#configuration)
to add to the `systemd_unit`'s `ExecStart`. Left `nil`, it will read from the
generated configuration file. If `generate_conf_file` is `false`, you can pass
your own configuration file and/or run-time options here.

### goiardi\_schob

Installs a precompiled binary of Schob, the Goiardi Shovey job client. Full
syntax with default values:

```
goiardi_schob 'schob' do
  instance_name resource_name # Defaults to the resource name, used to generate paths and sub-resources
  instance_path '/usr/sbin'   # Where to put the binary. /usr/sbin is FHS compliant, but /usr/local/sbin could be argued as better.
  version 'latest'            # must be either 'latest' or a git tag. Ignored if bin_url is set explicitely
  bin_url nil                 # Left nil, the url will be autoresolved based on `version`
  user resource_name          # Asset owner, defaults to the resource's name
  group resource_name         # Asset group, defaults to resource's name
  conf {}                     # Configuration entries to merge/override the defaults. USe `nil` to negate a default entirely
  whitelist { whitelist: { 'chef-client': 'chef-client', 'cinc-client': 'cinc-client' } } # See below
  server_public_key nil       # Required property, the contents of the Shovey server public key.
  manage_conf true            # Set to `false` if you prefer to manage the configuration file through other means
  manage_systemd_service true # Set to false if you're not n systemd or want to handle it through other means
  options nil                 # A string of CLI options for Schob. Defaults to reading everything from the configuration file
  serf_unit_name 'serf'       # Will be added as `name.service` to the systemd unit's "After" and "Requires". Ignored if `manage_systemd_service` is `false`
end
```

Leaving `version` at it's default of 'latest' will result in Schob
auto-updating with each client run. You probably don't want that happening in
prod o.O

`conf` is a hash of [Schob configuration
directives](https://github.com/ctdk/schob) that is merged in with some sane
defaults, intended to create a simple but sturdy installation. Default
directives can be overriden with either a new value or `nil` to have them
ommited entirely.

`whitelist` is a hash of valid commands that shovey can execute. There is no
detailed documentation avaialble at this time but the author provides [this
example](https://github.com/ctdk/schob/blob/master/test/whitelist.json) for
reference.

`options` is a string of [Schob CLI options](https://github.com/ctdk/schob) to
add to the `systemd_unit`'s `ExecStart`. Left `nil`, it will read from the
generated configuration file. If `generate_conf_file` is `false`, you can pass
your own configuration file and/or run-time options here.

### goiardi\_serf

This one is out-of-scope, but provided for convenience. It installs Hashicorp
Serf, which is required by Shovey jobs. Full syntax:

```
goiardi_serf 'serf' do
  instance_name resource_name  # Defaults to the resource name
  instance_path '/usr/sbin'    # Where to put the binary. /usr/sbin is FHS compliant, but /usr/local/sbin could be argued as better.
  version '0.8.2'              # Ignored if bin_url is set explicitely
  archive_url nil              # Left nil, the url will be autoresolved based on `version`
  archive_dir nil              # Where to put the zip file. If left nil, will go to the serf user's home
  user resource_name           # Asset owner, defaults to the resource's name
  group resource_name          # Asset group, defaults to resource's name
  manage_systemd_service false # Set to false if you're not on systemd or want to handle it through other means
  conf {}                      # Configuration entries to merge/override the defaults. USe `nil` to negate a default entirely
  options nil                  # A string of CLI options for Serf. Defaults to reading everything from the configuration
 file
end
```

`conf` is a hash of [Serf configuration
directives](https://www.serf.io/docs/agent/options.html) that is merged in with
some sane defaults, intended to create an installation tailored to Goiardi
Shovey jobs. Default directives can be overriden with either a new value or
`nil` to have them ommited entirely.

`options` is a string of [Serf CLI
options](https://www.serf.io/docs/agent/options.html) to add to the
`systemd_unit`'s `ExecStart`. Left `nil`, it will read from the generated
configuration file.

### Important notes on key handling

Giardi Shovey jobs require that you generate a cryptographic keypair. The
`goiardi_install` and `goiardi_schob` resources will both generate a keypair by
default but it is highly recommend to generate your own key out-of-band and
deploy it using secure mechanisms.

To use your own key with `goiardi_install`, put your key somewhere restricted
on disk and point to it with the `private_key_path` property. `goiardi_install`
will chown/chmod the key as needed.

Its best to have all references to your private key go through a secret manager
such as Hashicorp Vault, which is [natively supported by
Goiardi](https://goiardi.readthedocs.io/en/latest/features/secrets.html). The
resources supports configuring Vault through the `conf` properties.

Finally, the `goiardi_schob` resource requires the `goiardi_install` public key
to function. Being a public key, you can put it cleartext in your private
cookbook without much worry, or read it from a file on drive with ruby's
`IO.read` (or similar methods). Remember to use `lazy {}` to read the key if it
is created in the same run.

## Recipe examples

A minimal install, best used only for testing:

```
goiardi_install 'goiardi'
```

A more complete example, with explicit version, configuration overrides,
Postgres as a database backend and shovey enabled:

```
ze_conf = {
    # 'time-slew' is included by default, but we don't want it in our conf
    'time-slew': nil,
    # Override the default log level of 'warning' with 'fatal'
    'log-level': 'fatal',
    # Activate and configure postgresql support
    'use-postgresql': true,
    'postgresql-username': 'pguser',
    'postgresql-password': 'PleaseGetMeFromVault!',
    'postgresql-host': 'localhost',
    'postgresql-port': 5432,
    'postgresql-dbname': 'goiardi',
    'postgresql-ssl=mode': 'enable'
    # Activate Shovey. `sign-priv-key` should be automatically resolved so we omit it
    'use-serf': true,
    'use-shovey': true,
  }

goiardi_install 'goiardi' do
  version 'v.0.11.10'
  user 'my_custom_system_user'
  group 'my_custom_system_group'
  conf ze_conf # This could very well be a free-form attribute
end
```

Minimalist client node example:

```
remote_file '/etc/shovey.pub' do
  source 'http://webserver.myorg.tld/shovey_server.pub'
  mode '0644'
  action :create
end

goiardi_serf 'serf'

goiardi_schob 'schob' do
  goiardi_endpoint '127.0.0.1:4545'
  server_public_key { IO.read('/etc/shovey.pub') }
end
```

## Contributing

See CONTRIBUTING.md

## License

[Apache-2.0](https://opensource.org/licenses/Apache-2.0)

## Authors

The Cinc project

maintainers@cinc.io


CHANGELOG.md
# goiardi CHANGELOG

## 1.0.0

Initial release; Implements goiardi\_install and schob resources.


spec/unit/recipes/default_spec.rb
#
# Cookbook:: goiardi
# Spec:: default
#
# Copyright:: 2019, The Authors, All Rights Reserved.

require 'spec_helper'

describe 'goiardi::default' do
  context 'When all attributes are default, on Ubuntu 18.04' do
    # for a complete list of available platforms and versions see:
    # https://github.com/chefspec/fauxhai/blob/master/PLATFORMS.md
    platform 'ubuntu', '18.04'

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end

  context 'When all attributes are default, on CentOS 7' do
    # for a complete list of available platforms and versions see:
    # https://github.com/chefspec/fauxhai/blob/master/PLATFORMS.md
    platform 'centos', '7'

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


test/fixtures/policyfiles/remove.rb
# A name that describes what the system you're building with Chef does.
name 'test_goiardi_remove'

# Where to find external cookbooks:
default_source :supermarket

# run_list: chef-client will run these recipes in the order specified.
run_list 'test::remove'

# Specify a custom source for a single cookbook:
cookbook 'goiardi', path: '../../../'
cookbook 'test', path: '../cookbooks/test'


test/fixtures/policyfiles/default.rb
# A name that describes what the system you're building with Chef does.
name 'test_goiardi_default'

# Where to find external cookbooks:
default_source :supermarket

# run_list: chef-client will run these recipes in the order specified.
run_list 'goiardi::default', 'test::default'

# Specify a custom source for a single cookbook:
cookbook 'goiardi', path: '../../../'
cookbook 'test', path: '../cookbooks/test'


resources/serf.rb
# Installs Hashicorp Serf, technically out of scope for a Goiardi cookbook
# but is provided for convenience when installing Shovey. If you plan to
# use Serf for other things than Goiradi Shovey jobs I recommend installing
# it from a different cookbook.
property :instance_name, String, name_property: true
property :instance_path, String, default: '/usr/sbin'
property :version, String, default: '0.8.2'
property :archive_url, [String, NilClass], default: nil
property :archive_dir, [String, NilClass], default: nil
property :user, String, name_property: true
property :group, String, name_property: true
property :manage_systemd_service, [TrueClass, FalseClass], default: true
property :conf, Hash, default: {}
property :options, [String, NilClass], default: nil

default_action :install

action :install do
  group new_resource.group do
    comment "Generated by #{Chef::Dist::PRODUCT} for #{new_resource.instance_name}"
    append true
    system true
    action :create
  end

  user new_resource.user do
    comment "Generated by #{Chef::Dist::PRODUCT} for #{new_resource.instance_name}"
    gid new_resource.group
    shell '/sbin/nologin'
    home serf_home
    manage_home true
    system true
    action :create
  end

  remote_file "#{archive_directory}/#{new_resource.instance_name}.zip" do
    source new_resource.archive_url || derive_serf_download_url
    owner new_resource.user
    group new_resource.group
    mode '0755'
    notifies :extract, "archive_file[#{archive_directory}/#{new_resource.instance_name}.zip]", :immediately
  end

  archive_file "#{archive_directory}/#{new_resource.instance_name}.zip" do
    destination new_resource.instance_path
    overwrite true
    action :nothing
  end

  file bin_path do
    group new_resource.group
    owner new_resource.user
    mode '0755'
  end

  all_dirs.each do |dir|
    directory dir do
      owner new_resource.user
      group new_resource.group
      mode '0750'
      recursive true
      action :create
    end
  end

  file "#{conf_dir}/#{new_resource.instance_name}.json" do
    owner new_resource.user
    group new_resource.group
    mode '0644'
    content Chef::JSONCompat.to_json_pretty(merged_conf.compact)
    action :create
  end

  if new_resource.manage_systemd_service
    systemd_unit "#{new_resource.instance_name}.service" do
      content <<-EOU.gsub(/^\s+/, '')
      [Unit]
      Description=Hashicorp Serf service orchestrator
      After=network.target

      [Service]
      Type=exec
      PIDFile=/var/lock/#{new_resource.instance_name}.pid
      ExecStart=#{bin_path} agent #{serf_options}
      Restart=On-failure

      [Install]
      WantedBy=multi-user.target
      EOU
      action [:create, :enable, :start]
      subscribes :restart, "file[#{conf_dir}/#{new_resource.instance_name}.json]", :delayed
      subscribes :restart, "remote_file[#{bin_path}]", :delayed
    end
  end
end

action :remove do
  if new_resource.manage_systemd_service
    systemd_unit "#{new_resource.instance_name}.service" do
      action [:stop, :disable, :delete]
    end
  end

  file bin_path do
    action :delete
  end

  all_dirs.each do |dir|
    directory dir do
      recursive true
      action :delete
    end
  end

  user new_resource.user do
    action :remove
  end

  group new_resource.group do
    action :remove
  end
end

action_class do
  require 'chef/dist'
  include Goiardi::Helpers

  def serf_options
    return new_resource.options if new_resource.options
    new_resource.options ? new_resource.options : "-config-file #{conf_dir}/#{new_resource.instance_name}.json"
  end

  def conf_dir
    "/etc/#{new_resource.instance_name}"
  end

  def all_dirs
    dirs = [
      archive_directory,
      conf_dir,
    ]
    %w(queue_save_file whitelist).each do |file|
      dirs.push(::File.dirname(merged_conf[file.to_sym])) if merged_conf[file.to_sym]
    end
    clean_directories(dirs.compact.uniq, new_resource.instance_name)
  end

  def serf_home
    "/home/#{new_resource.user}"
  end

  def archive_directory
    new_resource.archive_dir || serf_home
  end

  def bin_path
    "#{new_resource.instance_path}/#{new_resource.instance_name}"
  end

  def derive_serf_download_url
    "https://releases.hashicorp.com/serf/#{new_resource.version}/#{serf_archive_name(node['os'], system_arch(node['kernel']['machine']), new_resource.version)}"
  end

  def merged_conf
    {
      node_name: node['fqdn'],
      log_level: 'warn',
    }.merge(new_resource.conf.to_h).compact
  end
end


resources/schob.rb
# Installs and configures schob, the Shovey client
# Shovey's like Chef Push Jobs, but based on Hashicorp's Serf
property :instance_name, String, name_property: true
property :instance_path, String, default: '/usr/sbin'
property :version, String, default: 'latest'
property :bin_url, [String, NilClass], default: nil
property :user, String, name_property: true
property :group, String, name_property: true
property :goiardi_endpoint, String, required: true
property :conf, Hash, default: {}
property :whitelist, Hash, default: { whitelist: { 'chef-client': 'chef-client', 'cinc-client': 'cinc-client' } }
property :server_public_key, String, required: true
property :manage_conf, [TrueClass, FalseClass], default: true
property :manage_systemd_service, [TrueClass, FalseClass], default: true
property :options, [String, NilClass], default: nil
property :serf_unit_name, String, default: 'serf'

default_action :install

action :install do
  group new_resource.group do
    comment "Generated by #{Chef::Dist::PRODUCT} for #{new_resource.instance_name}"
    append true
    system true
    action :create
  end

  user new_resource.user do
    comment "Generated by #{Chef::Dist::PRODUCT} for #{new_resource.instance_name}"
    gid new_resource.group
    shell '/sbin/nologin'
    system true
    action :create
  end

  all_dirs.each do |dir|
    directory dir do
      owner new_resource.user
      group new_resource.group
      mode '0750'
      recursive true
      action :create
    end
  end

  remote_file bin_path do
    # Using node attributes in resources is often considered poor form
    # But these are automatic attributes and I have faith in ohai!
    source new_resource.bin_url || release_by_platform_url('ctdk/schob', node['os'], system_arch, new_resource.version)
    owner new_resource.user
    group new_resource.group
    mode '0755'
  end

  # Schob won't start without it's own private key
  openssl_rsa_private_key "#{conf_dir}/#{new_resource.instance_name}.pem" do
    group new_resource.group
    owner new_resource.user
    sensitive true
  end

  openssl_rsa_public_key "#{conf_dir}/#{new_resource.instance_name}.pub" do
    group new_resource.group
    owner new_resource.user
    private_key_path "#{conf_dir}/#{new_resource.instance_name}.pem"
    sensitive true
  end

  # The server's pub key
  file "#{conf_dir}/shovey.pub" do
    owner new_resource.user
    group new_resource.group
    mode '0644'
    content new_resource.server_public_key
    action :create
  end

  file "#{conf_dir}/#{new_resource.instance_name}_whitelist.json" do
    owner new_resource.user
    group new_resource.group
    mode '0640'
    content Chef::JSONCompat.to_json_pretty(new_resource.whitelist)
    action :create
  end

  if new_resource.manage_conf
    template "/etc/#{new_resource.instance_name}/#{new_resource.instance_name}.conf" do
      source 'schob.conf.erb'
      cookbook 'goiardi'
      owner new_resource.user
      group new_resource.group
      mode '0644'
      variables(conf: merged_conf.compact)
      action :create
    end
  end

  if new_resource.manage_systemd_service
    systemd_unit "#{new_resource.instance_name}.service" do
      content <<-EOU.gsub(/^\s+/, '')
      [Unit]
      Description=Shovey client to receive jobs from a Goiardi server
      After=network.target #{new_resource.serf_unit_name}.service
      Requires=#{new_resource.serf_unit_name}.service

      [Service]
      Type=exec
      User=#{new_resource.user}
      PIDFile=/var/lock/#{new_resource.instance_name}.pid
      ExecStart=#{new_resource.instance_path}/#{new_resource.instance_name} #{schob_options}
      Restart=On-failure

      [Install]
      WantedBy=multi-user.target
      EOU
      action [:create, :enable, :start]
      subscribes :restart, "template[#{conf_dir}/#{new_resource.instance_name}.conf]", :delayed
      subscribes :restart, "remote_file[#{bin_path}]", :delayed
    end
  end
end

action :remove do
  if new_resource.manage_systemd_service
    systemd_unit "#{new_resource.instance_name}.service" do
      action [:stop, :disable, :delete]
    end
  end

  file bin_path do
    action :delete
  end

  file "#{conf_dir}/#{new_resource.instance_name}.conf" do
    action :delete
  end

  all_dirs.each do |dir|
    directory dir do
      recursive true
      action :delete
    end
  end

  user new_resource.user do
    action :remove
  end

  group new_resource.group do
    action :remove
  end
end

action_class do
  require 'chef/dist'
  # Generic helpers, those that do not rely directly on resource properties or ohai data
  include Goiardi::Helpers

  # Resource specific helper methods
  def schob_options
    new_resource.options ? new_resource.options : "-c #{conf_dir}/#{new_resource.instance_name}.conf"
  end

  def conf_dir
    "/etc/#{new_resource.instance_name}"
  end

  def all_dirs
    dirs = [
      conf_dir,
    ]
    %w(log-file key-file sign-pub-key whitelist queue-save-file).each do |f|
      dirs.push(::File.dirname(merged_conf[f.to_sym])) if merged_conf[f.to_sym]
    end
    clean_directories(dirs.compact.uniq, new_resource.instance_name)
  end

  def bin_path
    "#{new_resource.instance_path}/#{new_resource.instance_name}"
  end

  def system_arch
    node['kernel']['machine'] == 'x86_64' ? 'amd64' : node['kernel']['machine']
  end

  def merged_conf
    {
      'log-file': "/var/log/#{new_resource.instance_name}/#{new_resource.instance_name}.log",
      syslog: false,
      endpoint: new_resource.goiardi_endpoint,
      'node-name': node['fqdn'],
      'key-file': "#{conf_dir}/#{new_resource.instance_name}.pem",
      'sign-pub-key': "#{conf_dir}/shovey.pub",
      whitelist: "#{conf_dir}/#{new_resource.instance_name}_whitelist.json",
      'queue-save-file': "/var/lib/#{new_resource.instance_name}/queue_save_file",
    }.merge(new_resource.conf.to_h).compact
  end
end


libraries/helpers.rb
module Goiardi
  module Helpers
    require 'json'

    def release_by_platform_url(repo, platform, arch, version)
      asset = release_data(repo, version)['assets'].detect { |a| a['name'] == release_bin_name(repo, platform, arch, version) }
      raise 'Failed to locate asset' unless asset
      asset['browser_download_url']
    end

    def clean_directories(dirs, name)
      # crude but effective method to prevent chowning/deleting system directories
      dirs.reject { |d| !d.chomp('/').match(/#{name}/) }
    end

    def system_arch(arch)
      arch == 'x86_64' ? 'amd64' : arch
    end

    def serf_archive_name(platform, arch, version)
      ['serf', plain_version(version), platform, "#{arch}.zip"].join('_')
    end

    private

    def release_bin_name(repo, platform, arch, version)
      [repo.split('/').last, plain_version(expand_version(repo, version)), platform, arch].join('-')
    end

    def version_as_tag(repo, version)
      expanded = expand_version(repo, version)
      expanded.match(/^v/) ? expanded : "v#{expanded}"
    end

    def plain_version(version)
      version.start_with?('v') ? version.gsub(/^v/, '') : version
    end

    def expand_version(repo, version)
      if version == 'latest'
        return @latest_version if defined?(@latest_version)
        @latest_version = JSON.parse(Chef::HTTP.new('https://api.github.com').get("/repos/#{repo}/releases/latest"))['tag_name']
      else
        version
      end
    end

    def release_data(repo, version)
      return @release_data if defined?(@release_data)
      @release_data = JSON.parse(Chef::HTTP.new('https://api.github.com').get("/repos/#{repo}/releases/tags/#{version_as_tag(repo, version)}"))
    end
  end
end


resources/install.rb
# This resource installs the goiardi bin and all its' prerequisites
property :instance_name, String, name_property: true
property :instance_path, String, default: '/usr/sbin'
property :version, String, default: 'latest'
property :bin_url, [String, NilClass], default: nil
property :user, String, name_property: true
property :group, String, name_property: true
property :conf, Hash, default: {}
property :private_key_path, [String, NilClass], default: nil
property :manage_conf, [TrueClass, FalseClass], default: true
property :manage_systemd_service, [TrueClass, FalseClass], default: true
property :options, [String, NilClass], default: nil

default_action :install

action :install do
  group new_resource.group do
    comment "Generated by #{Chef::Dist::PRODUCT} for #{new_resource.instance_name}"
    append true
    system true
    action :create
  end

  user new_resource.user do
    comment "Generated by #{Chef::Dist::PRODUCT} for #{new_resource.instance_name}"
    gid new_resource.group
    shell '/sbin/nologin'
    system true
    action :create
  end

  remote_file bin_path do
    # Using node attributes in resources is often considered poor form
    # But these are automatic attributes and I have faith in ohai!
    source new_resource.bin_url || release_by_platform_url('ctdk/goiardi', node['os'], system_arch(node['kernel']['machine']), new_resource.version)
    owner new_resource.user
    group new_resource.group
    mode '0755'
  end

  all_dirs.each do |dir|
    directory dir do
      owner new_resource.user
      group new_resource.group
      mode '0750'
      recursive true
      action :create
    end
  end

  # If the user is providing his own private key, chown it to the correct user
  if merged_conf.key?(:'use-shovey')
    if new_resource.private_key_path
      file new_resource.private_key_path do
        user new_resource.user
        group new_resource.group
        mode '0600'
        action :create
      end

    else
      openssl_rsa_private_key "#{conf_dir}/#{new_resource.instance_name}.pem" do
        group new_resource.group
        owner new_resource.user
        sensitive true
      end

      openssl_rsa_public_key "#{conf_dir}/#{new_resource.instance_name}.pub" do
        group new_resource.group
        owner new_resource.user
        private_key_path "#{conf_dir}/#{new_resource.instance_name}.pem"
        sensitive true
      end
    end
  end

  if new_resource.manage_conf
    template "#{conf_dir}/#{new_resource.instance_name}.conf" do
      source 'goiardi.conf.erb'
      cookbook 'goiardi'
      owner new_resource.user
      group new_resource.group
      mode '0644'
      variables(conf: merged_conf.compact)
      action :create
      notifies :restart, "systemd_unit[#{new_resource.instance_name}.service]", :delayed if new_resource.manage_systemd_service
    end
  end

  if new_resource.manage_systemd_service
    systemd_unit "#{new_resource.instance_name}.service" do
      content <<-EOU.gsub(/^\s+/, '')
      [Unit]
      Description=Goiardi, a go implementation of the Chef Server
      After=network.target

      [Service]
      Type=exec
      User=#{new_resource.user}
      PIDFile=/var/lock/#{new_resource.instance_name}.pid
      ExecStart=#{new_resource.instance_path}/#{new_resource.instance_name} #{goiardi_options}
      Restart=On-failure

      [Install]
      WantedBy=multi-user.target
      EOU
      action [:create, :enable, :start]
      subscribes :restart, "remote_file[#{bin_path}]", :delayed
    end
  end
end

action :remove do
  if new_resource.manage_systemd_service
    systemd_unit "#{new_resource.instance_name}.service" do
      action [:stop, :disable, :delete]
    end
  end

  file bin_path do
    action :delete
  end

  all_dirs.each do |dir|
    directory dir do
      recursive true
      action :delete
    end
  end

  user new_resource.user do
    action :remove
  end

  group new_resource.group do
    action :remove
  end
end

action_class do
  require 'chef/dist'
  # Generic helpers, those that do not rely directly on resource properties or ohai data
  include Goiardi::Helpers

  # Resource specific helper methods
  def goiardi_options
    new_resource.options ? new_resource.options : "-c #{merged_conf[:'conf-root']}/#{new_resource.instance_name}.conf"
  end

  def all_dirs
    dirs = [
      merged_conf[:'local-filestore-dir'],
      merged_conf[:'conf-root'],
    ]
    # we're only going to operate on folders that match the instance name
    # to avoid recursively chowning/deleting important directories
    %w(log-file index-file data-file).each do |f|
      dirs.push(::File.dirname(merged_conf[f.to_sym])) if merged_conf[f.to_sym]
    end
    clean_directories(dirs.compact.uniq, new_resource.instance_name)
  end

  def bin_path
    "#{new_resource.instance_path}/#{new_resource.instance_name}"
  end

  def conf_dir
    "/etc/#{new_resource.instance_name}"
  end

  def merged_conf
    base = {
      ipaddress: '0.0.0.0',
      port: 4545,
      'index-file': "/var/lib/#{new_resource.instance_name}/#{new_resource.instance_name}-index.bin",
      'data-file': "/var/lib/#{new_resource.instance_name}/#{new_resource.instance_name}-data.bin",
      'log-file': "/var/log/#{new_resource.instance_name}/#{new_resource.instance_name}.log",
      'local-filestore-dir': "/var/lib/#{new_resource.instance_name}/#{new_resource.instance_name}-lfs",
      'freeze-interval': 120,
      'time-slew': '15m',
      'conf-root': conf_dir,
      'use-auth': true,
      'use-ssl': false,
      syslog: false,
      'log-level': 'warning',
      'disable-webui': false,
      'log-events': true,
      'log-event-keep': 1000,
    }
    base[:'sign-priv-key'] = if new_resource.private_key_path
                               new_resource.private_key_path
                             elsif new_resource.conf.to_h.key?('use-shovey')
                               "#{conf_dir}/#{new_resource.instance_name}.pem"
                             end
    # Note the use of to_h, it ensures that if a user passes attributes in directly as `conf`
    # that it will be treated as a plain hash instead of a Chef::ImmutableMash
    base.merge(new_resource.conf.to_h).compact
  end
end
